### ISSUE TITLE
Middleware to handle retries for SQLSTATE value of '40001' (PostgreSQL, CockroachDB)

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#11183]
> Messages with the Postgres error code 40001 indicate that a transaction failed because it conflicted with another concurrent or recent transaction accessing the same data. The transaction needs to be retried by the client.

More context: https://www.cockroachlabs.com/docs/stable/error-handling-and-troubleshooting.html#transaction-retry-errors
And for PostgreSQL: https://www.postgresql.org/docs/current/transaction-iso.html#:~:text=It%20is%20important%20that%20an,back%20to%20prevent%20serialization%20anomalies

How to trigger via: https://github.com/cockroachlabs/example-app-node-prisma/pull/1#discussion_r785223589

> It should be possible to deterministically generate a real error of this type by beginning one transaction, committing a conflicting transaction, and attempting to commit the first transaction.

**Comment by Jolg42** [comment]
Just found this recommendation
>In production, we recommend implementing client-side transaction retries for all database operations. 
Pointing to this
https://www.cockroachlabs.com/docs/stable/transactions#client-side-intervention

**Comment by doowb** [comment]
Hi, I thought I'd try out CockroachDB and just ran across this `SqlState(E40001)` error when running a seed script that we've been using successfully with Postgres.

Most specifically, the CockroachDB error message starts with:

```
restart transaction: TransactionRetryWithProtoRefreshError: WriteTooOldError
```

We aren't specifically using transactions in the places that this is happening, so is the prisma client or engine specifying a transaction that may be conflicting? This seems like it should be something that's handled by prisma's connection to the db, or at least documented better (I didn't see any mention in the CochroachDB section of the Prisma docs)

I'm going back to using Postgres, but I thought I'd add that I ran into this issue when trying to switch (the linked comment above indicates that you, *Prisma*, haven't seen this issue much)

**Comment by janpio** [comment]
Note that this kind of error should now actually be caught by Prisma, and bubbled up as a known error that is easier to handle in middlewares and similar: https://github.com/prisma/prisma-engines/blob/4dac7298d82af85904827c171dae069c7334841d/libs/user-facing-errors/src/query_engine.rs#L337-L338

```
code = "P2034",
message = "Transaction failed due to a write conflict or a deadlock. Please retry your transaction"
```

**Comment by jhaynie** [comment]
we are trying to figure out how to implement the retry as it seems like any session related TX settings seem to not be handled correctly.  We've tried monkeypatching `$transaction` to implement the advanced client side retry logic from Cockroach DB documentation.  We can't seem to figure out the right path to make this work.

Here's a quick gist of what I've tried:

https://gist.github.com/jhaynie/485360624a2ae19bd5b3a4903037d8c8

It seems like the underlying prisma engine will automatically issue a ROLLBACK when receiving an error (which makes sense) so we can't seem to call `ROLLBACK TO SAVEPOINT` (will raise another exception).  Also, on success, when we call `RELEASE SAVEPOINT cockroach_restart` we get an error (something about transaction already released).

So we can't seem to do this in middleware (we tried that too but it seems like calling any of the required additional TX parameters don't work) and we can't seem to monkeypatch in this behavior.

What would be the recommended approach to handling this?

**Comment by Jolg42** [comment]
Seems that this could also be useful for MongoDB `Transaction failed due to a write conflict or a deadlock. Please retry your transaction`
See https://github.com/prisma/prisma/issues/12814

**Comment by janpio** [comment]
@jhaynie (Every time I see you pop up here, your name rings a bell and I have to hover over your profile to jog my brain to remember Appcelerator - good old times 😆 Wrote my thesis about cross platform mobile frameworks back in the day...)

Am I correct in understanding that you are trying to implement https://www.cockroachlabs.com/docs/v22.1/advanced-client-side-transaction-retries and can not make it work with our interactive transactions?

**Comment by janpio** [comment]
Great you could make it work. This is a _very_ good illustration why we indeed need to offer a better solution here.
(Talking to colleagues internally right now how that might be solvable via https://github.com/prisma/prisma/issues/15074)

**Comment by rafbgarcia** [comment]
hey @janpio, is there a better way of achieving the transaction retry with the release of the client extensions?

I looked around but couldn't find an example (I'm new to Prisma so maybe there is and I didn't find it)

**Comment by SevInf** [comment]
@rafbgarcia would something along this lines work for you:

```ts
const xprisma = prisma.$extends({
  client: {
    $transaction(callback) {
      return prisma.$transaction(async (tx) => {
        await tx.$exectuteRawUnsafe("SAVEPOINT cockroach_restart");

        for (let i = 0; i < 5; i++) {
          try {
            const result = await callback(tx);
            await tx.$executeRawUnsafe("RELEASE SAVEPOINT cockroach_restart");
            return result;
          } catch (error) {
            if (error.code === "P2034") {
              await tx.$executeRawUnsafe(
                "ROLLBACK TO SAVEPOINT cockroach_restart"
              );
              continue;
            }
            throw error;
          }
        }
      });
    },
  },
});
```

Haven't tested it personally yet, but I think that would be rough direction where I'd go with implementing it via extension

**Comment by rafbgarcia** [comment]
@SevInf thank you that's very helpful!

**Comment by ssijak** [comment]
Any changes regarding this in later Prisma versions?

**Comment by jleni** [comment]
any updates here? this is blocking us :(

**Comment by iserdmi** [comment]
Hello! I solved my problems like this with this solution: https://github.com/prisma/prisma-client-extensions/tree/main/retry-transactions

TLDR:
```
$ npm i exponential-backoff
```

```

import { backOff, IBackOffOptions } from "exponential-backoff";
import { Prisma, PrismaClient } from "@prisma/client";

function RetryTransactions(options?: Partial<IBackOffOptions>) {
  return Prisma.defineExtension((prisma) =>
    prisma.$extends({
      client: {
        $transaction(...args: any) {
          return backOff(() => prisma.$transaction.apply(prisma, args), {
            retry: (e) => {
              // Retry the transaction only if the error was due to a write conflict or deadlock
              // See: https://www.prisma.io/docs/reference/api-reference/error-reference#p2034
              return e.code === "P2034";
            },
            ...options,
          });
        },
      } as { $transaction: typeof prisma["$transaction"] },
    })
  );
}

const prisma = new PrismaClient().$extends(
  RetryTransactions({
    jitter: "full",
    numOfAttempts: 5,
  })
);
```
